libtransact/src/scheduler/serial/core.rs

```diff
@@ -156,92 +170,150 @@ impl SchedulerCore {
+                    })
+                }
+                TransactionExecutionResult::Invalid(_) => {
+                    panic!("Should not already have an invalid result");
+                }
+            })
+            .collect();
+
+        self.txn_results
+            .push(TransactionExecutionResult::Invalid(invalid_result));
+
+        // Invalidate all unexecuted transactions in the batch
+        self.txn_results.append(
+            &mut self
+                .txn_queue
+                .drain(..)
+                .map(|txn| {
+                    TransactionExecutionResult::Invalid(InvalidTransactionResult {
+                        transaction_id: txn.header_signature().into(),
+                        error_message: "batch invalid".into(),
+                        error_data: vec![],
+                    })
+                })
+                .collect(),
+        );
+    }
+
+    fn send_batch_result(&mut self) {
+        let batch = self
+            .current_batch
+            .take()
+            .expect("received execution result but no current batch is executing");
+
+        let mut results = vec![];
+        std::mem::swap(&mut results, &mut self.txn_results);
+
+        let batch_result = BatchExecutionResult { batch, results };
+
+        let shared = self
+            .shared_lock
+            .lock()
+            .expect("scheduler shared lock is poisoned");
+        match shared.result_callback() {
+            Some(callback) => {
+                callback(Some(batch_result));
+            }
+            None => {
+                warn!(
+                    "dropped batch execution result: {}",
+                    batch_result.batch.batch().header_signature()
+                );
+            }
+        }
+    }
+
     fn run(&mut self) -> Result<(), CoreError> {
         loop {
             match self.rx.recv() {
                 Ok(CoreMessage::BatchAdded) => {
                     self.try_schedule_next()?;
                 }
                 Ok(CoreMessage::ExecutionResult(task_notification)) => {
-                    let batch = match self.current_batch.take() {
-                        Some(batch) => batch,
-                        None => {
-                            panic!("received execution result but no current batch is executing")
-                        }
-                    };
-
-                    let results = match task_notification {
+                    match task_notification {
                         ExecutionTaskCompletionNotification::Valid(context_id) => {
+                            let transaction = self.current_txn.take().expect(
+                                "received execution result but no current txn is executing",
+                            );
                             self.previous_context = Some(context_id);
-                            vec![TransactionExecutionResult::Valid(
+                            self.txn_results.push(TransactionExecutionResult::Valid(
```

Comment:
Should we check that the notification matches the current transaction here, so if the executor is misbehaving we catch it?

Reply:
`ExecutionTaskCompletionNotification::Valid` would need to be updated to include the transaction ID; should I do that?

Reply:
It probably should, since you will need that for the parallel scheduler - no guarantees about order there.

Reply:
Updated the notification to include the transaction ID and added a check (for both valid and invalid notifications) to make sure the notification is for the expected transaction.